Allow Shims to replace Hive execs

mythrocks opened the pull request:
Distro-specific shims are currently allowed to customize operator/exec/expr replacements in most cases. However, replacements made specifically for exprs/execs from `org.apache.spark.sql.hive` (i.e. the ones handled by `HiveProvider`) do not succeed.

For instance, the `HiveTableScanExec` is currently provided by `HiveProviderImpl`. When a Spark distribution's shim attempts to block/override this exec's replacement, it does not succeed, because the `HiveProviderImpl` changes are applied *last* in `GpuOverrides`.

This change puts the Spark shim replacement rules at the end, so that the shim has the final say.

mythrocks commented:
Build

mythrocks commented:
The build failure on 3.4 seems to be unrelated to this current change:
```
[ERROR] [Error] /home/mithunr/workspace/dev/spark-plugin/2/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOptimizedCreateHiveTableAsSelectCommand.scala:31: object OptimizedCreateHiveTableAsSelectCommand is not a member of package org.apache.spark.sql.hive.execution
[ERROR] [Error] /home/mithunr/workspace/dev/spark-plugin/2/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOptimizedCreateHiveTableAsSelectCommand.scala:118: not found: type OptimizedCreateHiveTableAsSelectCommand
[ERROR] [Error] /home/mithunr/workspace/dev/spark-plugin/2/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOptimizedCreateHiveTableAsSelectCommand.scala:152: not found: type OptimizedCreateHiveTableAsSelectCommand
[ERROR] [Error] /home/mithunr/workspace/dev/spark-plugin/2/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOptimizedCreateHiveTableAsSelectCommand.scala:148: not found: type OptimizedCreateHiveTableAsSelectCommand
[ERROR] [Error] /home/mithunr/workspace/dev/spark-plugin/2/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOptimizedCreateHiveTableAsSelectCommand.scala:19: Unused import
[ERROR] [Error] /home/mithunr/workspace/dev/spark-plugin/2/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOptimizedCreateHiveTableAsSelectCommand.scala:31: Unused import
[ERROR] [Error] /home/mithunr/workspace/dev/spark-plugin/2/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3625: type arguments [org.apache.spark.sql.execution.command.CreateDataSourceTableAsSelectCommand] do not conform to method dataWriteCmd's type parameter bounds [INPUT <: org.apache.spark.sql.execution.command.DataWritingCommand]
[ERROR] [Error] /home/mithunr/workspace/dev/spark-plugin/2/sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/HiveProviderImpl.scala:32: object OptimizedCreateHiveTableAsSelectCommand is not a member of package org.apache.spark.sql.hive.execution
[ERROR] [Error] /home/mithunr/workspace/dev/spark-plugin/2/sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/HiveProviderImpl.scala:43: not found: type OptimizedCreateHiveTableAsSelectCommand
[ERROR] [Error] /home/mithunr/workspace/dev/spark-plugin/2/sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/HiveProviderImpl.scala:32: Unused import
```
Apache Spark has removed `OptimizedCreateHiveTableAsSelect` on `master`. I should check if this is being addressed elsewhere.

mythrocks commented:
@jlowe: Thank you for the suggestion earlier, to reorder the expressions/execs in `GpuOverrides.scala`. That would have worked, were it not for the package structure.

Unfortunately, `HiveTableScanExec` cannot be replaced directly from the Shims (or `GpuOverrides`), because it is package private to `org.apache.spark.sql.hive`. It needs to be replaced from the `HiveProvider`. It is easier to pipe the choice of `HiveProvider` through the shim, and replace it there.

mythrocks commented:
Build

jlowe commented:
> Unfortunately, HiveTableScanExec cannot be replaced directly from the Shims (or GpuOverrides), because it is package private to org.apache.spark.sql.hive

That does not prevent shims from overriding HiveTableScanExec.  Shims can, and do, have sources that are in different packages.  See the stuff in org.apache.spark instead of com.nvidia.spark under the varous shims, as one example.  Since this override meta would be essentially private to the shim, it can be located in any package it wants to be.

jlowe commented on sql-plugin/src/main/311until320-nondb/scala/com/nvidia/spark/rapids/shims/Spark31XShims.scala:
```diff
@@ -41,6 +41,7 @@ import org.apache.spark.sql.execution.exchange.{BroadcastExchangeExec, ReusedExc
 import org.apache.spark.sql.execution.joins._
```
2023 copyrights here and every other file in the PR.

mythrocks replied:
Thank you. Fixed.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:
```diff
@@ -4048,7 +4048,7 @@ object GpuOverrides extends Logging {
 
   lazy val execs: Map[Class[_ <: SparkPlan], ExecRule[_ <: SparkPlan]] =
-    commonExecs ++ SparkShimImpl.getExecs ++ GpuHiveOverrides.execs
+    commonExecs ++ SparkShimImpl.getExecs ++ SparkShimImpl.getHiveProvider.getExecs
```
IMO it's weird to ask the shim for all of its execs and then also, separately, ask it for the hive execs.  If the shim wants to put Hive in its execs, it can do so directly.  To be clear it's not a must-fix for me, but I do find it odd to have two separate APIs for getting execs from a shim.

mythrocks commented:
Build

mythrocks commented:
I have modified this patch based on @jlowe's suggestion. I agree that this works a little better:
1. The change is now minuscule.
2. The Shim developer is not exposed to `HiveProvider`. They may replace Hive execs as they see fit. Less coupling all around.


I have changed the brief and description to match the change.

An example of how a shim (say for CDH) may modify the replacement rules for a Hive exec (say `HiveTableScanExec`) is provided [here](https://github.com/NVIDIA/spark-rapids/commit/2d2943b59acedbc8dfbf8f8e1e15757be5ed8c70).

mythrocks commented:
Thank you for the review, @revans2, @jlowe. I have merged this change.
